### PR TITLE
Add s390x for ubuntu and centos artifacts

### PR DESCRIPTION
**Pull request by aditi-sharma-1**
**What this PR does / why we need it**:
This pull request adds s390x architecture in Containerdisks for Ubuntu and CentOS artifacts. These changes ensure compatibility and resolve build issues encountered on the s390x architecture. 

**Release note**:

```release-note
Add s390x artifacts
```

Signed-off-by: Aditi Sharma [Aditi.Sharma@ibm.com](mailto:Aditi.Sharma@ibm.com)

### CODE REVIEW

File: Makefile
Comment (aditi-sharma-1) [reply Makefile:1]: Hi @0xFelix, Thanks for the review. Sure, will make these changes in a separate commit.

Comment (aditi-sharma-1) [reply Makefile:1]: Done. Thank you.

File: cmd/medius/common/registry.go
Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go:31]: Hi @0xFelix , Sure, looking it.

Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go:31]: Hi @0xFelix , 
Yes, I have checked, and it's working. I am attaching the logs below for your reference. Thank you. 
[s390x-build.txt](https://github.com/user-attachments/files/17382483/s390x-build.txt)

Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go:31]: When I first executed `bin/medius images push --target-registry=localhost:5000 --dry-run=false --insecure-skip-tls --force --focus=ubuntu:22.04`, I encountered the following error:

`error introspecting image "quay.io/containerdisks/ubuntu:22.04": Error parsing manifest for image: choosing image instance: no image found in manifest list for architecture "s390x", variant "", OS "linux" `

Since I was trying to build it locally on s390x, I commented out the section of code in `cmd/medius/images/push.go` (from [line 291](https://github.com/kubevirt/containerdisks/blob/main/cmd/medius/images/push.go#L291) to [line 304](https://github.com/kubevirt/containerdisks/blob/main/cmd/medius/images/push.go#L304)) that checks the quay.io manifest and modified the function to return true. I have included the git diff output in the logs for your reference. After making this change, the execution was successful. I have also done a `docker pull `to verify the images (`ubuntu:22.04 and centos-stream:9`) created for the s390x architecture.

Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go:31]: Please let me know if you have any suggestions or specific requirements for this implementation. I appreciate your guidance! Thank you.

Comment (0xFelix) [reply cmd/medius/common/registry.go:31]: > I commented out the section of code [...] that checks the quay.io manifest

There is a `--source-registry` flag to change the source registry. That might have helped, or for next time. 

Did you run `medius images verify` to verify the images? I'm interested to see if the built images were able to boot successfully.

Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go:31]: Hi @0xFelix ,
I am trying to execute `medius images verify` but facing some errors. I'm currently investigating the issue. I will keep you updated about the progress.

Thank you for your understanding!

Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go:31]: Hi @0xFelix, @lyarwood, 

I have successfully verified the locally built images for s390x after making the following adjustments and would like to kindly request your consideration in adding s390x support:

1. I used [this kubevirtci fork](https://github.com/chandramerla/kubevirtci) for starting the cluster with the `make cluster-up` command. This PR is raised by one of our team members to introduce s390x support in the kubevirtci repository.
2. I modified pipeline.sh, replacing the CirrOS focus image with ubuntu:22.04.

Please let me know if you have any questions. Thank you very much for your consideration!

File: artifacts/fedora/fedora.go
Comment (lyarwood) [inline artifacts/fedora/fedora.go:198]: This doesn't exist yet in common-instancetypes, can you post a PR?

File: cmd/medius/common/registry.go
Comment (lyarwood) [reply cmd/medius/common/registry.go:31]: @aditi-sharma-1 apologies myself and @0xFelix have been out on PTO.

Are there any efforts to merge the changes of that fork into core `kubevirtci`?

Are there also any efforts to provide `s390x` hosts as CI runners that this project could use to verify these images?

Without this we can't merge these changes.

Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go:31]: Hi @lyarwood,
My teammate is actively working on the kubevirtci PR, and its progress can be tracked here: https://github.com/kubevirt/kubevirtci/pull/1316.

Regarding the availability of s390x machines, as mentioned by one of my teammates earlier (https://github.com/kubevirt/containerdisks/pull/182#issuecomment-2312558107), it will take some time to provide these resources to host CI runners. 

I have executed the pipeline.sh file with the locally built Ubuntu image, and it is working well. If you would like, I can share the logs for your review. 

Could you please let me know which additional test cases or files need to be executed for the CI process? I will also try to run these locally for the images to ensure comprehensive testing. 

Thank you!

File: artifacts/fedora/fedora.go
Comment (aditi-sharma-1) [reply artifacts/fedora/fedora.go:198]: Hi @lyarwood @0xFelix ,

I will move this commit to a new PR and continue the discussions there. I will keep this PR focused on the Ubuntu and CentOS artifacts only.

Since the build process for Fedora differs from that of Ubuntu and CentOS, could you please advise on how you would suggest adding preferences here for a new architecture? As far as I know, the common instance types also depend on the containerdisks repository during their testing. 

Looking forward to your reply. Thanks in advance!

**Comment by LakshmiRavichandran1** [comment]
Hi @lyarwood  @0xFelix, thank you for your review and comments on the PR.

I co-work with @aditi-sharma-1 on the s390x efforts.

I take this opportunity to explain to you the importance and urgency of this PR and why we need it merged on the s390x side at the earliest.

The downstream tests (OCP virt - tier 0/1 tests), has a tight dependency expecting the container disks images for fedora, centos, ubuntu to be available from `quay.io`.
We are ticking on the timeline to unblock the downstream tests and to have a clean run next week.

kindly, is it possible to consider the urgency of the PR and a possibility to merge the changes with the confidence that we got the testing covered from our local s390x verification?
(Aditi can give you the logs again, please refer https://github.com/kubevirt/containerdisks/pull/203#discussion_r1801649950).
Meanwhile, we are also motivated to provide you a s390x CI environment but that would take some time.

so, is it possible to push the changes forward with so called label '`untested-on-CI`' so we mark it now and re-visit later with motivation?

Also, for the related changes for s390x/fedora in this PR, we have covered the local s390x verification for it inclusively from the common-instance types. so, these changes are also good to get piggy backed.

Please let me know your thoughts.
Thank you.

**Comment by 0xFelix** [comment]
I have no problem adding the containerdisks, but I would ask you to remove the environment variables for instancetypes and preferences from them for now, as they don't exist yet.

File: artifacts/fedora/fedora.go
Comment (0xFelix) [reply artifacts/fedora/fedora.go:198]: You are right that there is a chicken and egg problem. I would start by adding the containerdisks without environment variables for instancetypes/preferences. The second step would be to create suitable instancetypes and preferences and the last step would be to add them to the containerdisks once they are verified working.

Adding the containerdisks without environment variables as a first step should get you unblocked quickly.

**Review by 0xFelix — COMMENTED** [review]
Did you drop the unit tests for centosstream and ubuntu by accident? Also, did you drop support for Fedora on purpose?

**Comment by aditi-sharma-1** [comment]
Hi @0xFelix, 
I have added the unit tests for the CentOS and Ubuntu artifacts.

I am currently working on the Fedora artifact and will raise a separate PR for it soon. This PR focuses on the Ubuntu and CentOS artifacts.

I’ve also attached the logs for building and verification of the Ubuntu and CentOS artifacts on s390x for your reference.

Thank you! 

[containerdisks_building&verification_onS390x.txt](https://github.com/user-attachments/files/17676466/containerdisks_building.verification_onS390x.txt)

File: artifacts/ubuntu/ubuntu_test.go
Comment (0xFelix) [inline artifacts/ubuntu/ubuntu_test.go:95]: nit: empty line

File: cmd/medius/common/registry.go
Comment (0xFelix) [inline cmd/medius/common/registry.go:161]: nit: empty line

**Review by 0xFelix — COMMENTED** [review]
/approve

Please remove the empty lines, otherwise looks good to me.

**Comment by 0xFelix** [comment]
/approve cancel

Can you please check the CI failures? They are related.

File: artifacts/ubuntu/ubuntu_test.go
Comment (aditi-sharma-1) [reply artifacts/ubuntu/ubuntu_test.go]: Done. Thank you.

File: cmd/medius/common/registry.go
Comment (aditi-sharma-1) [reply cmd/medius/common/registry.go]: Updated the file. Thank you.

**Review by 0xFelix — COMMENTED** [review]
The remaining failure should be due to the image not existing on quay yet.

File: pkg/architecture/arch.go
Comment (aditi-sharma-1) [reply pkg/architecture/arch.go:5]: Hi @0xFelix ,

Previously, I was using the `case "s390x": return "s390x"` directly inside `GetImageArchitecture`. I added it as a constant after the CI showed an error related to `golangci-lint` on Friday. The error message was:

```
pkg/architecture/arch.go:11:7: string s390x has 2 occurrences, make it a constant (goconst)
	case "s390x":
	     ^
```


Sure, I can revert to the previous format and add the comment `//nolint:goconst` to avoid the linter error about the repeated string. Please let me know if you would prefer this approach. 
Thank you!

Comment (aditi-sharma-1) [reply pkg/architecture/arch.go]: I have updated the file. Thanks :)

**Review by 0xFelix — COMMENTED** [review]
/approve

@lyarwood PTAL

**Review by lyarwood — APPROVED** [review]
/lgtm

**Comment by lyarwood** [comment]
/hold

```
time="2024-11-11T14:39:35Z" level=error msg="error introspecting image \"quay.io/containerdisks/ubuntu:24.04\": Error parsing manifest for image: choosing image instance: no image found in manifest list for architecture \"s390x\", variant \"\", OS \"linux\"" name=ubuntu version=24.04
time="2024-11-11T14:39:35Z" level=error msg="error introspecting image \"quay.io/containerdisks/ubuntu:22.04\": Error parsing manifest for image: choosing image instance: no image found in manifest list for architecture \"s390x\", variant \"\", OS \"linux\"" name=ubuntu version=22.04
```

I don't recall hitting this with the arm64 images? Looking now.

**Comment by lyarwood** [comment]
Reproduces locally FWIW....

```
$ make medius
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o bin/medius kubevirt.io/containerdisks/cmd/medius
$ bin/medius images push --target-registry=localhost:5000 --dry-run=false --insecure-skip-tls --focus=ubuntu:22.04
INFO[0005] Latest containerdisk checksum: "55c687a9a242fab7b0ec89ac69f9def77696c4e160e6f640879a0b0031a08318"  name=ubuntu version=22.04
INFO[0008] Latest containerdisk checksum: "2edb369b85141fbeff6c87c7d92e08f315ec236f639235e35429e4930d98e2de"  name=ubuntu version=22.04
ERRO[0009] error introspecting image "quay.io/containerdisks/ubuntu:22.04": Error parsing manifest for image: choosing image instance: no image found in manifest list for architecture "s390x", variant "", OS "linux"  name=ubuntu version=22.04
INFO[0009] Writing results file                         
FATA[0009] error introspecting image "quay.io/containerdisks/ubuntu:22.04": Error parsing manifest for image: choosing image instance: no image found in manifest list for architecture "s390x", variant "", OS "linux" 
```

**Comment by 0xFelix** [comment]
> The error is coming from `github.com/containers/image/v5`, quick and dirty workaround below works for me for now

@lyarwood  Thats more of a bug on our side. I knew there was handling for unknown tags in medius (hence my comment), but I did not know that unknown archs have to be handled separately. Thanks for addressing it!

Thanks for you contribution @aditi-sharma-1! :)

**Comment by lyarwood** [comment]
> > The error is coming from `github.com/containers/image/v5`, quick and dirty workaround below works for me for now
> 
> @lyarwood Thats more of a bug on our side. I knew there was handling for unknown tags in medius (hence my comment), but I did not know that unknown archs have to be handled separately. Thanks for addressing it!

Yeah I think that handling is new however as I don't recall it being an issue with the `arm64` disks, still good to get this landed finally.

**Pull request merged**
